### ISSUE TITLE
[Inductor][Compile]torch.compile remembers the old weight for evaluation if grad is disabled

### ISSUE THREAD

**Issue by youkaichao** [pytorch/pytorch#108863]
### 🐛 Describe the bug

Code to reproduce the bug is at the [gist](https://gist.github.com/youkaichao/7ed49dcb55b2e66dfd841b1a9b0bfeff).

The reason:

When grad is disabled and conv-bn modules are in eval mode, the `fuse_conv_bn` optimization is turned on:

https://github.com/pytorch/pytorch/blob/09f7cb0eaf02d69ab4c788b03345198f5c869718/torch/_inductor/fx_passes/pre_grad.py#L97

The optimization fuses parameters directly, creating a computation graph with **a new conv** whose parameters are computed from the conv and bn parameters **at the time of compiling the model**.

Consider the procedure of eval-train-eval:
1. When we evaluate the model at the first time, the newly created conv in the compiled graph remembers the old parameter values;
2. Then when we train the model, the parameters' value changed, but the newly created conv is not updated.
3. When we evaluate the model again, the guarding conditions for the eval graph remains satisfied, so this doesn't trigger re-compilation. The eval graph constructed in Step 1 is used directly, with the obsolete values.

The bug affects all the training loop who uses `torch.compile` and evaluates the model with `torch.no_grad()`.

This bug is very difficult to find and check, as only consecutive conv-bn values in a whole neural network are affected. The eval graph runs with some updated parameters and some obselete parameters. Only when I dive into the details and shrink the network to have only one pair of conv-bn, does the bug appear clearly.

A possible fix is to fuse computation rather than parameters for conv-bn, as described in https://github.com/pytorch/pytorch/pull/108757 : 

- When we find a pair of conv-bn usage, as long as the bn is in eval mode, we change the computation of `normalize (weight conv input)` to `(normalize (weight)) conv input`. These two are mathematically equivalent, and does not touch the parameters.
- If users explicit specify something like `TORCHINDUCTOR_FREEZING=1`, we know the parameters cannot be updated. Then the constant folding pass can automatically fold the `normalize (weight)` into a baked `weight`.

This solution is sound, and can also deal with training conv-bn when bn is kept in eval mode.

### Versions

I'm testing on `2.1.0.dev20230904`, as well as nightly `2.2` built from main.

**Comment by youkaichao** [comment]
cc @jansel @Chillee @eellison

**Comment by jansel** [comment]
@eellison is this expected behavior when freezing is enabled?  I assume this only happens when freezing is on.

**Comment by youkaichao** [comment]
This would be expected if freezing is on. There are some overlapping conditions:
- 1. no grad
- 2. module in eval state
- 3. freezing (not update parameters any more)

While typically 3 ==> 1 & 2, the reverse is not True: 1 & 2 cannot imply 3. This is the root cause of the bug.

**Comment by jansel** [comment]
Sounds like we just need to add an extra check before applying freezing.

**Comment by youkaichao** [comment]
Yeah, disabling the optimization if we detect it is not freezing is always a doable action, it is sound, although this means users get less optimization :(

**Comment by youkaichao** [comment]
My point is that we can "fuse the computation" rather than "fuse the parameters".

Fusing parameters only works if users explicitly specify the freezing option, while fusing computation without changing the parameters applies to training/evaluation/inference.

The following graph illustrates what I mean by changing the computation:

<img width="475" alt="image" src="https://github.com/pytorch/pytorch/assets/23236638/776dab18-611f-43b8-afb2-5e2fad8cbe39">

If we view consecutive conv and bn in eval mode as a function, it is a pure function without any side effect:

```python
def conv_bn_computation(x, conv_weight, bn):
    x = conv(x, conv_weight)
    return bn(x)
```

And it is safe to use a mathematically equivalent but faster implementation:

```python
def fused_conv_bn_computation(x, conv_weight, bn):
    conv_weight = bn(conv_weight) # broadcast the bn operation properly
    x = conv(x, conv_weight)
    return x
```

These two functions have exactly the same set of input parameters and features, and the same output features, therefore they are indistinguishable and interchangeable, while the latter is faster and more memory efficient.

My recent goal is to improve the conv-bn fusion in pytorch from parameter fusion to computation fusion. To achieve that goal, I'm willing to take all the trouble in learning and integrating with the infrastructure of pytorch :)

**Comment by eellison** [comment]
This isn't freezing that's happening but a cpu pass. @XiaobingSuper what happened with revert of https://github.com/pytorch/pytorch/pull/106686 ?

**Comment by XiaobingSuper** [comment]
> This isn't freezing that's happening but a cpu pass. @XiaobingSuper what happened with revert of #106686 ?
The reason is that there is a big accuracy loss(I test some CV models on imagenet dataset) when do conv+bn folding at bfloat16 parameters: when we do the folding at the post AOTAutograd step, we will insert to_dtype for convolution before freezing path, the freezing pass will mark the bfloat16 convolution's weight as a constant parameter at first, then do conv+bn folding(conv parameters is bfloat16), even we using float32 as accumulation dtype for the conv+bn folding, there still has a precision loss, so I hope the conv+bn folding happened before AOTAutograd and using the float32 parameters do the folding.
